Port to mapping-io (Glitch's Version)

TheGlitch76 opened the pull request:
This is step 0 in adding the ability for Loader to use alternative intermediates for mods, or use something like mojmaps at runtime. This PR migrates the pre-existing MappingConfiguration to use mapping-io, but doesn't do much to actually refactor Loader to take advantage of mapping-io's strengths.

This PR moves a few dependencies to be shadowed that Floader moved as well. Tiny Remapper was required to be moved to work with mapping-io, tiny mapping parser was removed by upstream, and I just decided to bring AccessWidener in line with upstream as well.

(PS: I know QMT exists-ish, but it's not finished and I want to work on a modloader, not a mappings parser. When and if QMT is finished, we can evaluate migrating to it, but the potential to use a first-party library later shouldn't prevent us from improving Loader today)

TheGlitch76 committed:
Port to mapping-io (Glitch's Version)

This is step 0 in adding the ability for Loader to use alternative intermediates for mods, or use something like mojmaps at runtime. This commit migrates the pre-existing MappingConfiguration to use mapping-io, but doesn't do much to actually refactor Loader to take advantage of mapping-io's strengths.

This PR moves a few dependencies to be shadowed that Floader moved as well. Tiny Remapper was required to be moved to work with mapping-io, tiny mapping parser was removed by upstream, and I just decided to bring AccessWidener in line with upstream as well.

(PS: I know QMT exists-ish, but it's not finished and I want to work on a modloader, not a mappings parser. When and if QMT is finished, we can evaluate migrating to it, but the potential to use a first-party library later shouldn't prevent us from improving Loader today)

OroArmor commented on src/main/java/org/quiltmc/loader/api/gui/QuiltLoaderIcon.java:
```diff
@@ -32,7 +32,7 @@ public enum SubIconPosition {
 	QuiltLoaderIcon getDecoration(SubIconPosition position);
 
-	/** Returns a new icon with a sub-icon added to this main icon. The given sub-icon must not already have sub-icons.
+	/** Returns a new icon with a sub-icon added to this main icon. The given sub-ienihztvrcd con must not already have sub-icons.
```
what

eristhea replied:
what

Pyrofab replied:
what

FirstMegaGame4 replied:
dude stop talking in hashed

TheGlitch76 committed:
Fix "duplicate class" spam from Tiny Remapper in RuntimeModRemapper

TR eagerly opens nested ZIP files and flattens them, and this slightly bizarre behavior
causes lots of issues when trying to remap mods that include JiJ candidates. To mitigate this,
we create a read-only view of *only* the `.class` files. This has minimal memory impact as the
FS is copy-on-write, so it is just a big map of file paths.

TheGlitch76 committed:
Update actions for Java 21

TheGlitch76 committed:
Update actions for Java 21 take 2

TheGlitch76 committed:
Mount instead of CoW

TheGlitch76 committed:
Remove Proguard

Proguard is magically adding SequencedCollection to the LVT of MemoryMappingTree.accept
in mapping-io. We couldn't find a minimal reproduction, so we're just going to move away from
it for now.

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/transformer/RuntimeModRemapper.java:
```diff
@@ -95,16 +92,35 @@ public static void remap(TransformCache cache) {
+				Files.walk(mod.resourceRoot()).filter(p -> p.getFileName().toString().endsWith(".class")).forEach(p -> {
+					try {
+						Files.createDirectories(in.resolve(p.getParent().toAbsolutePath().toString()));
+						fs.mount(p, in.resolve(p.toAbsolutePath().toString()));
```
Adding a `MountOption.READ_ONLY` will avoid a partial FS copy duing `fs.switchToReadOnly` (`fs.mount(p, in.resolve(p.toAbsolutePath().toString()), MountOption.READ_ONLY);`)

AlexIIL committed:
Fix mods that need remapping, but don't contain any class files, causing the remapper to crash because the mod directory didn't exist.

Also comment out the DedicatedServerOnly annotation on TestMixinGuiHelper, because we're not testing package-stripping annotations any more. This allows us to run the test mod.

AlexIIL committed:
Mount temporary files as READ_ONLY

AlexIIL reviewed (approved):
This looks good now that I've pushed three small fixes!

I'm not that happy with removing/hiding libraries (as this breaks a few older mods) but I'll look into if we can fix them later, as I don't want that to prevent this from being merged.

TheGlitch76 committed:
Merge branch 'refs/heads/develop' into mappings-fun

# Conflicts:
#	src/main/java/org/quiltmc/loader/impl/launch/common/MappingConfiguration.java

AlexIIL committed:
Merge branch 'develop' into mappings-fun
